### ISSUE TITLE
RFE: Add `MaxVersion=` to sysupdate.d

### ISSUE THREAD

**Issue by septatrix** [systemd/systemd#30695]
### Component

other

### Is your feature request related to a problem? Please describe

Currently sysupdate does not allow specifying a max version to update to, however, it is not uncommon to restrict the versions one should be able to update to without intermediate steps. E.g. most Linux distros only allow upgrading from the previous version but upgrading multiple major version is often unsupported as that significantly increases the amount of potential problems. While it is possible to encode this as part of the file URL (e.g. upload files under `.../v1/foobar_1.0.img`) this does not extend to more complicated scenarios easily (e.g. update at most 2 major versions).

### Describe the solution you'd like

The `[Transfer]` section of the sysupdate definitions should support a `MaxVersion=` property similar to the existing `MinVersion=`. Furthermore, it should be possible to specify whether the bound is inclusive or exclusive to allow for scenarios like `MaxVersion= < 2.0.0` analogous to [`ConditionKernelVersion=`](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#ConditionKernelVersion=).
For backwards compatibility the default should be `<=`, unlike `ConditionKernelVersion=` which uses `$=`. It is like also sufficient to only support `<, <=, >=, >` as the rest does not make much sense for these bounds.

### Describe alternatives you've considered

It would also be possible to deprecate `MinVersion=` and instead use `VersionBound=`. With the support for specifying the comparison operator this would allow specifying both the upper and lower bound. If one would have thought of this from the start that would likely have been the ideal solution. While sysupdate is still considered experimental (AFAIK) this would not be the end of the world but I can understand that one wants to avoid this.

PS: [Most recent mailing list entry](https://lists.freedesktop.org/archives/systemd-devel/2024-January/049894.html) about this topic

### The systemd version you checked that didn't have the feature you are asking for

255

**Comment by AdrianVovk** [comment]
I think the use cases you describe are solved by https://github.com/systemd/systemd/pull/33706 (which closes https://github.com/systemd/systemd/issues/33345)

---

For example, let's say you have v3 of a distro that does major version upgrades. During v4's development, they stabilize what the sysupdate transfers will look like and ship those files in `/usr/lib/sysupdate-next.d` in the next security patch to v3. Eventually, v4 is ready for release, so they publish it onto their servers. Let's say the user chooses not to upgrade.

Vendor is now developing v5. They push v5's transfer definitions to v4's sysupdate-next.d directory. Once v5 is released, they can test a v3 -> v5 upgrade, and push an update to v3 systems that swaps out the sysupdate-next directory from v4 to v5 (clearly the user wasn't interested in v4!). If v3 -> v5 doesn't work all that well, then the v3 system would be forced to upgrade through v4 first. Let's say the user chooses not to upgrade again, and sticks to v3.

Vendor is now developing v6. Same thing happens, except vendor is no longer pushing any new transfer definitions to v3. Eventually v3 users go EOL, so they either need to do a clean install of v6, or upgrade through v4 (and possibly v5) to get to v6.

---

Or are you talking about limiting the number of versions to jump forward in a single stream? I.e. in a situation like ChromeOS, where there's only one update stream and the moment v4 comes out v3 is immediately EOL and no longer receives its own updates.

**Comment by septatrix** [comment]
> Or are you talking about limiting the number of versions to jump forward in a single stream? I.e. in a situation like ChromeOS, where there's only one update stream and the moment v4 comes out v3 is immediately EOL and no longer receives its own updates.

Yes that is more my use-case though there is certainly an overlap of functionality covered by `--next` or `MaxVersion=`. The former would mean that a normal sysupdate invocation still defaults to a minor update. The latter however seems to brings much more flexibility with regards to how one can specify version bounds and seems to be more capable. It would also be a nice symmetry to the existing `MinVersion=` instead of a new concept. I see little advantage of `--next` over a `MaxVersion=` bound. The functionality covered by `--next` seems to be easily covered by `--definitions=...` which also has the advantage that an image could provide several streams (eg. next-major, beta, next-lts, etc.)

**Comment by AdrianVovk** [comment]
Well MaxVersion= doesn't actually cover the use-case for `--next`, since there's no way to allow the user to upgrade to the next major release (it's impossible to enumerate what the "next" major release is), nor can it handle distributions changing their transfer definitions during major version upgrades. `--definitions=` doesn't work either since it's impossible to discover its availability programmatically, and it doesn't correctly handle administrator overrides in /etc.

Though you bring up a good point about other streams. `--next` can pretty trivially be converted into `--stream=` to make the concept more generic. I'll do that, thanks.

But back to the topic on hand, in single-stream situations IMO MaxVersion= absolutely makes sense. Contributions welcome! (though I'd suggest waiting for the entire stack of sysupdate PRs to land first, see the dependency list of the --next PR I linked earlier)

**Comment by quentin9696** [comment]
Hi ! 

Any update on that one ?

Thanks
